### ISSUE TITLE
bug(android): Can't backspace after pasting string with encoded URI components

### ISSUE THREAD

**Issue by ermshiperete** [keymanapp/keyman#11182]
### Describe the bug

I have the link below in the address bar in the Chrome browser and the IP at the end of the string. When I press Backspace (with EuroLatin keyboard active) nothing happens.

https://cdn.retailads.net/tc.php?t=151238C2287900209D&deepurl=https%3A%2F%2Fa.nonstoppartner.net%2Fa%2F%3Fi%3Dclick%26l%3Dde%26client%3Daldi%26camp%3Ddeep%26nw%3Dread%26deepid%3D101012640&subid=1712319X100149708X67035X29711X86818X181XtextlinkXyahl6byko6csX

https://github.com/keymanapp/keyman/assets/181336/05b4f5b7-b9fe-446b-a5df-ef29f942e40d

### Keyman apps

- [X] Keyman for Android
- [ ] Keyman for iPhone and iPad
- [ ] Keyman for Linux
- [ ] Keyman for macOS
- [ ] Keyman for Windows
- [ ] Keyman Developer
- [ ] KeymanWeb
- [ ] Other - give details at bottom of form

### Keyman version

17.0.301-beta

### Operating system

Android 12

### Device

realme X3 SuperZoom (RMX2086)

### Browser

Chrome

### Keyboard name

EuroLatin (SIL)

**Comment by darcywong00** [comment]
I reprod.
1. With Keyman as the system keyboard, 
2. Select Chrome bar (empty)
3. paste the text into the Chrome bar
4. attempt to backspace (does nothing)

I suspect KeymanWeb isn't updated and still treats the context as empty.

**Comment by darcywong00** [comment]
ok, it's tied to the length of the context
I can't backspace at the end of this
`https://cdn.retailads.net/tc.php?t=151238C2287900209D&deepurl=https%3A%2F%2Fa.nonstoppartner.net%2Fa%2F%3Fi%3Dclick%26l%3Dde%26client%3Daldi%26camp%3Ddeep%26nw%3Dread%26deepid%3D101012640&subid=1712319X100149708X67035X29711`

but I can from
`https://cdn.retailads.net/tc.php?t=151238C2287900209D&deepurl=https%3A%2F%2Fa.nonstoppartner.net%2Fa%2F%3Fi%3Dclick%26l%3Dde%26client%3Daldi%26camp%3Ddeep%26nw%3Dread%26deepid%3D101012640&subid=1712319X100149708X67035X2971'

**Comment by darcywong00** [comment]
More debugging
> 3. paste the text into the Chrome bar

I see in Android Studio this causes SystemKeyboard.onUpdateSelection() to get called

 > 4. attempt to backspace (does nothing)

I set a breakpoint in the Chrome debugger of android-host.js and I can see when hitting backspace, insertText() doesn't get triggered from KeymanWeb.
Hitting a different key does trigger insertText().

@jahorton - is there an internal KeymanWeb transform buffer that should get updated when pasting context? (afaict, pasting text does **not** trigger KMManager.resetContext()

**Comment by jahorton** [comment]
> @jahorton - is there an internal KeymanWeb transform buffer that should get updated when pasting context? (afaict, pasting text does **not** trigger KMManager.resetContext()

Yes.  The buffer can be force-updated via a call to `keyman.context.saveState()`.

https://github.com/keymanapp/keyman/blob/93dc41e2f5e1892f0de3fa821f44da8fbd144355/web/src/app/webview/src/contextManager.ts#L52

Note that said function is called automatically when text is set via `setText`:

https://github.com/keymanapp/keyman/blob/93dc41e2f5e1892f0de3fa821f44da8fbd144355/web/src/app/webview/src/contextManager.ts#L96-L107

I'd expect this to have been in the pipeline already, though - surely the pasted text is passed along to the internal KeymanWeb engine?

> I set a breakpoint in the Chrome debugger of android-host.js and I can see when hitting backspace, `insertText()` doesn't get triggered from KeymanWeb.

That shouldn't be the case; if text is selected (selection range width > 0), a backspace should trigger an `insertText` call.  So... something's going on here, for sure.

**Comment by jahorton** [comment]
Oh, _this_ again...

![image](https://github.com/keymanapp/keyman/assets/25213402/05a717c6-acd3-4367-8912-046f3c4b699c)

The internal Web engine's selection range is improperly specified - the link is 222 characters long, but the Web engine believes that the caret is at index 256.  The embedded context uses the cached selection-range values here:

https://github.com/keymanapp/keyman/blob/93dc41e2f5e1892f0de3fa821f44da8fbd144355/common/web/keyboard-processor/src/text/outputTarget.ts#L420-L430

As a result, backspaces don't actually affect anything aside from selection-range values until the selection-range resynchronizes.

**Comment by jahorton** [comment]
Interestingly, the values _are_ correct on the Android side - it appears that they get transformed when passed through to Web.

```
//What Android sees and specifies: 
https://cdn.retailads.net/tc.php?t=151238C2287900209D&deepurl=https%3A%2F%2Fa.nonstoppartner.net%2Fa%2F%3Fi%3Dclick%26l%3Dde%26client%3Daldi%26camp%3Ddeep%26nw%3Dread%26deepid%3D101012640&subid=1712319X100149708X67035X29711X86818X181XtextlinkXyahl6byko6csX

// What _Web_ receives:
https://cdn.retailads.net/tc.php?t=151238C2287900209D&deepurl=https://a.nonstoppartner.net/a/?i=click&l=de&client=aldi&camp=deep&nw=read&deepid=101012640&subid=1712319X100149708X67035X29711X86818X181XtextlinkXyahl6byko6csX
```

Since the text has changed, the selection-range values are no longer correct.

**Comment by jahorton** [comment]
Ah, and here's the culprit:  our JS queuing setup doesn't bother with encoding the string.

https://github.com/keymanapp/keyman/blob/93dc41e2f5e1892f0de3fa821f44da8fbd144355/android/KMEA/app/src/main/java/com/keyman/engine/KMKeyboard.java#L331-L341

As the JS call is passed in URL form, the encoded URI patterns are auto-decoded before KeymanWeb gets a chance to touch them.

**Comment by darcywong00** [comment]
Some more testing of pasting & deleting
`%52` ▶️ first two backspaces do nothing, then it starts deleting
`?52` ▶️ deletes fine
`/52` ▶️ deletes fine

**Comment by darcywong00** [comment]
Fixed by #11206
